### ISSUE TITLE
Searching for term with 2 words created unnecessary list of results

### ISSUE THREAD

**Issue by ADepic** [osmandapp/OsmAnd#6643]
This is still a problem. If you search 2 words (it normally applies for addresses now) it seems to give all the addresses containing just one of those 2 words, then the POI with those 2 words.

E.g

Search for "xyz abc"

Results:

xyz tyu
xyz dfg
xyz agf
xyz dhs
xyz abc

This is very annoying. In my opinion That should be removed. Search results should only be shown if both words are present.

This is a duplicate of #5921 , but the commit there never solved it

**Comment by vshcherb** [comment]
We need concrete search terms in order to test and solve it. In your case the sorting of results might be an issue

**Comment by ADepic** [comment]
Anything on this? This search mechanic is very frustrating.

This is the quickfix in my opinion:

When searching for "car wash", the app currently returns all results in the nearby area containing "car" and "wash". This should be changed. The app should ONLY return results with both "car" and "wash".

There are very few cases I can think of where having returning every single result with that word can be a good thing. If you search "car" "wash", you don't also want "car maintenance" or "bike wash" or, "car shop" or "2nd hand car shop"

As for bus stops (or railway station or any POI), for example with "Deptford Bridge" bus sop, if the user searches "Deptford Bridge" and "Deptford Bridge Bus Stop", both should return the correct result. When searching for POIs, the POI should be indexed as " " - So for example: "Deptford Bridge Bus Stop", "Paddington Railway Station". What is displayed as the results will be "Deptford Bridge" along with "Bus Stop" in a smaller font below it.

This small fix should make the search 100,000x better and it is very easy to implement

**Comment by vshcherb** [comment]
We don't consider it is very easy unfortunately so we need to attack specific examples. For car wash the result is decent:
![image](https://user-images.githubusercontent.com/1042025/67275500-0dde8e00-f4c3-11e9-971e-c009404e45d7.png)

**Comment by Friptick** [comment]
Completely agree with OP. This is unexpected behavior and deeply annoying. Unfortunately the search functionality is generally bad in this otherwise excellent app. Look at all the open issues, usually tagged "Nice to have" even though most are real bugs. The quick fix proposed here, i.e. changing an OR to an AND, would make search instantly better. Why is it difficult?

**Comment by sonora** [comment]
How about we sort the exact AND matches to the top, while not excluding the OR matches?

An issue with excluding them is if you e.g. typed 'Ambrosius St.' but the database contains 'Ambrosius Street', or you typed Saint Ambrosius' and the database contained 'St. Ambrosius', you would course the app for finding zero search hits no matter how unique your second search term 'Ambrosius' may be for the area.

**Comment by Friptick** [comment]
Yep exactly. That would be an instant improvement. And surely the way one expects search to work in the first place.

**Comment by jfburke3** [comment]
@sonora I wholeheartedly support that idea. The search function has long been an annoyance.

**Comment by sonora** [comment]
Yes, offline searching in my mind is likely the biggest weakness of OsmAnd and should be on our high urgency list to improve.

Needless to say, my reasoning is that search is a base functionality for any mapping or navigation app, and a self explanatory search interface which simply performs is key, and would camouflage all other complications and complexities we may have in the app to cater for fewer but much advanced users.

**Comment by oxxm** [comment]
Thanks for tackling this issue again.

As mentioned in #5921, how about showing only the **AND** results on the map? Consider the example of "the royal bank of scotland" in greater london.

**Comment by ADepic** [comment]
@vshcherb
What do you think?

**Comment by vshcherb** [comment]
My take on that is slightly changing, I think our biggest problem is that we don't properly support synonyms where Google search is using any bit of information and is creating a unique hashcode for each object, so it works for terms like "car wash", "car repair" = "car maintenance" giving exactly same result), similar to "bar" = "cocktails" = "drinks", "st" = "saint" ... we don't support it and we also don't have automated procedure to create it cause we neither parsing websites nor we have semantical language parser.
